File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/StoryViewModel.kt
package com.kissangram.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kissangram.model.Story
import com.kissangram.model.UserStories
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.FirestoreStoryRepository
import com.kissangram.usecase.GetStoryBarUseCase
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.delay
import kotlinx.coroutines.Job

class StoryViewModel(
    application: Application,
    initialUserId: String
) : AndroidViewModel(application) {

    private val authRepository = AndroidAuthRepository(
        context = application,
        activity = null,
        preferencesRepository = AndroidPreferencesRepository(application)
    )
    private val storyRepository = FirestoreStoryRepository(authRepository = authRepository)
    private val getStoryBarUseCase = GetStoryBarUseCase(storyRepository)

    private val _uiState = MutableStateFlow(StoryUiState())
    val uiState: StateFlow<StoryUiState> = _uiState.asStateFlow()

    private var autoAdvanceJob: Job? = null
    private val STORY_DURATION_MS = 5000L // 5 seconds per story

    init {
        loadStories(initialUserId)
    }

    fun loadStories(initialUserId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                val allUserStories = getStoryBarUseCase()
                Log.d(TAG, "loadStories: loaded ${allUserStories.size} user stories")
                
                if (allUserStories.isEmpty()) {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = "No stories available"
                    )
                    return@launch
                }

                // Find initial user index
                val initialIndex = allUserStories.indexOfFirst { it.userId == initialUserId }
                val startIndex = if (initialIndex >= 0) initialIndex else 0

                _uiState.value = _uiState.value.copy(
                    userStories = allUserStories,
                    currentUserIndex = startIndex,
                    currentStoryIndex = 0,
                    isLoading = false
                )

                // Start auto-advance for first story
                startAutoAdvance()
                markCurrentStoryAsViewed()
            } catch (e: Exception) {
                Log.e(TAG, "loadStories: failed", e)
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to load stories"
                )
            }
        }
    }

    fun nextStory() {
        val currentState = _uiState.value
        val currentUserStories = currentState.getCurrentUserStories() ?: return
        
        if (currentState.currentStoryIndex < currentUserStories.stories.size - 1) {
            // Move to next story in current user
            _uiState.value = currentState.copy(currentStoryIndex = currentState.currentStoryIndex + 1)
            startAutoAdvance()
            markCurrentStoryAsViewed()
        } else {
            // Move to next user
            nextUser()
        }
    }

    fun previousStory() {
        val currentState = _uiState.value
        val currentUserStories = currentState.getCurrentUserStories() ?: return
        
        if (currentState.currentStoryIndex > 0) {
            // Move to previous story in current user
            _uiState.value = currentState.copy(currentStoryIndex = currentState.currentStoryIndex - 1)
            startAutoAdvance()
            markCurrentStoryAsViewed()
        } else {
            // Move to previous user
            previousUser()
        }
    }

    fun nextUser() {
        val currentState = _uiState.value
        if (currentState.currentUserIndex < currentState.userStories.size - 1) {
            _uiState.value = currentState.copy(
                currentUserIndex = currentState.currentUserIndex + 1,
                currentStoryIndex = 0
            )
            startAutoAdvance()
            markCurrentStoryAsViewed()
        }
    }

    fun previousUser() {
        val currentState = _uiState.value
        if (currentState.currentUserIndex > 0) {
            val previousUserStories = currentState.userStories[currentState.currentUserIndex - 1]
            _uiState.value = currentState.copy(
                currentUserIndex = currentState.currentUserIndex - 1,
                currentStoryIndex = previousUserStories.stories.size - 1 // Start at last story
            )
            startAutoAdvance()
            markCurrentStoryAsViewed()
        }
    }

    fun pauseAutoAdvance() {
        autoAdvanceJob?.cancel()
        autoAdvanceJob = null
    }

    fun resumeAutoAdvance() {
        startAutoAdvance()
    }

    private fun startAutoAdvance() {
        autoAdvanceJob?.cancel()
        autoAdvanceJob = viewModelScope.launch {
            delay(STORY_DURATION_MS)
            nextStory()
        }
    }

    private fun markCurrentStoryAsViewed() {
        val currentState = _uiState.value
        val currentStory = currentState.getCurrentStory() ?: return
        
        if (!currentStory.isViewedByMe) {
            viewModelScope.launch {
                try {
                    storyRepository.markStoryAsViewed(currentStory.id)
                    // Update local state
                    val updatedUserStories = currentState.userStories.toMutableList()
                    val updatedUserStory = updatedUserStories[currentState.currentUserIndex]
                    val updatedStories = updatedUserStory.stories.toMutableList()
                    val storyIndex = updatedStories.indexOfFirst { it.id == currentStory.id }
                    if (storyIndex >= 0) {
                        updatedStories[storyIndex] = currentStory.copy(isViewedByMe = true)
                        updatedUserStories[currentState.currentUserIndex] = updatedUserStory.copy(
                            stories = updatedStories,
                            hasUnviewedStories = updatedStories.any { !it.isViewedByMe }
                        )
                        _uiState.value = currentState.copy(userStories = updatedUserStories)
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "markCurrentStoryAsViewed: failed", e)
                }
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        autoAdvanceJob?.cancel()
    }

    companion object {
        private const val TAG = "StoryViewModel"
    }
}

data class StoryUiState(
    val userStories: List<UserStories> = emptyList(),
    val currentUserIndex: Int = 0,
    val currentStoryIndex: Int = 0,
    val isLoading: Boolean = false,
    val error: String? = null
) {
    fun getCurrentUserStories(): UserStories? {
        return if (currentUserIndex in userStories.indices) {
            userStories[currentUserIndex]
        } else {
            null
        }
    }

    fun getCurrentStory(): Story? {
        val userStories = getCurrentUserStories() ?: return null
        return if (currentStoryIndex in userStories.stories.indices) {
            userStories.stories[currentStoryIndex]
        } else {
            null
        }
    }
}

File: iosApp/iosApp/viewmodel/StoryViewModel.swift
import Foundation
import os.log
import Shared

@MainActor
class StoryViewModel: ObservableObject {
    
    private static let log = Logger(subsystem: "com.kissangram", category: "StoryViewModel")
    private let prefs = IOSPreferencesRepository()
    private let authRepository: AuthRepository
    private let storyRepository: StoryRepository
    private let getStoryBarUseCase: GetStoryBarUseCase
    
    @Published var userStories: [UserStories] = []
    @Published var currentUserIndex: Int = 0
    @Published var currentStoryIndex: Int = 0
    @Published var isLoading: Bool = false
    @Published var error: String? = nil
    @Published var allStoriesFinished: Bool = false
    
    private var autoAdvanceTask: Task<Void, Never>? = nil
    private let STORY_DURATION_MS: UInt64 = 5_000_000_000 // 5 seconds in nanoseconds
    
    init(initialUserId: String) {
        self.authRepository = IOSAuthRepository(preferencesRepository: prefs)
        let userRepository = FirestoreUserRepository(authRepository: authRepository)
        let followRepository = IOSFollowRepository(authRepository: authRepository, userRepository: userRepository)
        self.storyRepository = FirestoreStoryRepository(authRepository: authRepository, followRepository: followRepository)
        self.getStoryBarUseCase = GetStoryBarUseCase(storyRepository: storyRepository)
        
        Task {
            await loadStories(initialUserId: initialUserId)
        }
    }
    
    func loadStories(initialUserId: String) async {
        isLoading = true
        error = nil
        Self.log.debug("loadStories: start, initialUserId=\(initialUserId)")
        
        do {
            let allUserStories = try await storyRepository.getStoryBar()
            Self.log.debug("loadStories: loaded \(allUserStories.count) user stories")
            
            if allUserStories.isEmpty {
                isLoading = false
                error = "No stories available"
                return
            }
            
            // Find initial user index
            let initialIndex = allUserStories.firstIndex { $0.userId == initialUserId } ?? 0
            
            self.userStories = allUserStories
            self.currentUserIndex = initialIndex
            self.currentStoryIndex = 0
            self.isLoading = false
            
            // Start auto-advance for first story
            startAutoAdvance()
            await markCurrentStoryAsViewed()
        } catch {
            Self.log.error("loadStories: failed: \(error.localizedDescription)")
            isLoading = false
            self.error = error.localizedDescription
        }
    }
    
    func nextStory() -> Bool {
        guard let currentUserStories = getCurrentUserStories() else { return false }
        
        if currentStoryIndex < currentUserStories.stories.count - 1 {
            // Move to next story in current user
            currentStoryIndex += 1
            startAutoAdvance()
            Task {
                await markCurrentStoryAsViewed()
            }
            return true
        } else {
            // Move to next user
            return nextUser()
        }
    }
    
    func previousStory() {
        guard let currentUserStories = getCurrentUserStories() else { return }
        
        if currentStoryIndex > 0 {
            // Move to previous story in current user
            currentStoryIndex -= 1
            startAutoAdvance()
            Task {
                await markCurrentStoryAsViewed()
            }
        } else {
            // Move to previous user
            previousUser()
        }
    }
    
    func nextUser() -> Bool {
        if currentUserIndex < userStories.count - 1 {
            currentUserIndex += 1
            currentStoryIndex = 0
            startAutoAdvance()
            Task {
                await markCurrentStoryAsViewed()
            }
            return true
        } else {
            // All stories finished
            allStoriesFinished = true
            return false
        }
    }
    
    func previousUser() {
        if currentUserIndex > 0 {
            let previousUserStories = userStories[currentUserIndex - 1]
            currentUserIndex -= 1
            currentStoryIndex = previousUserStories.stories.count - 1 // Start at last story
            startAutoAdvance()
            Task {
                await markCurrentStoryAsViewed()
            }
        }
    }
    
    func pauseAutoAdvance() {
        autoAdvanceTask?.cancel()
        autoAdvanceTask = nil
    }
    
    func resumeAutoAdvance() {
        startAutoAdvance()
    }
    
    private func startAutoAdvance() {
        autoAdvanceTask?.cancel()
        autoAdvanceTask = Task {
            try? await Task.sleep(nanoseconds: STORY_DURATION_MS)
            if !Task.isCancelled {
                let hasMore = nextStory()
                if !hasMore {
                    allStoriesFinished = true
                }
            }
        }
    }
    
    private func markCurrentStoryAsViewed() async {
        guard let currentStory = getCurrentStory() else { return }
        
        if !currentStory.isViewedByMe {
            do {
                try await storyRepository.markStoryAsViewed(storyId: currentStory.id)
                // Update local state
                if let userIndex = userStories.firstIndex(where: { $0.userId == getCurrentUserStories()?.userId }),
                   let storyIndex = userStories[userIndex].stories.firstIndex(where: { $0.id == currentStory.id }) {
                    var updatedStories = userStories[userIndex].stories
                    updatedStories[storyIndex] = Story(
                        id: currentStory.id,
                        authorId: currentStory.authorId,
                        authorName: currentStory.authorName,
                        authorUsername: currentStory.authorUsername,
                        authorProfileImageUrl: currentStory.authorProfileImageUrl,
                        authorRole: currentStory.authorRole,
                        authorVerificationStatus: currentStory.authorVerificationStatus,
                        media: currentStory.media,
                        textOverlay: currentStory.textOverlay,
                        locationName: currentStory.locationName,
                        visibility: currentStory.visibility,
                        viewsCount: currentStory.viewsCount,
                        likesCount: currentStory.likesCount,
                        isViewedByMe: true,
                        isLikedByMe: currentStory.isLikedByMe,
                        createdAt: currentStory.createdAt,
                        expiresAt: currentStory.expiresAt
                    )
                    userStories[userIndex] = UserStories(
                        userId: userStories[userIndex].userId,
                        userName: userStories[userIndex].userName,
                        userProfileImageUrl: userStories[userIndex].userProfileImageUrl,
                        userRole: userStories[userIndex].userRole,
                        userVerificationStatus: userStories[userIndex].userVerificationStatus,
                        stories: updatedStories,
                        hasUnviewedStories: updatedStories.contains { !$0.isViewedByMe },
                        latestStoryTime: userStories[userIndex].latestStoryTime
                    )
                }
            } catch {
                Self.log.error("markCurrentStoryAsViewed: failed: \(error.localizedDescription)")
            }
        }
    }
    
    func getCurrentUserStories() -> UserStories? {
        guard currentUserIndex < userStories.count else { return nil }
        return userStories[currentUserIndex]
    }
    
    func getCurrentStory() -> Story? {
        guard let userStories = getCurrentUserStories(),
              currentStoryIndex < userStories.stories.count else { return nil }
        return userStories.stories[currentStoryIndex]
    }
    
    deinit {
        autoAdvanceTask?.cancel()
    }
}
